### ISSUE TITLE
I need python caffe version of wingloss

### ISSUE THREAD

**Issue by alexcruz0202** [FengZhenhua/Wing-Loss#16]
thank you for your job. I am interested in your work. Can you provide caffe version code of wingloss?

**Comment by eyiztan** [comment]
same here@_@  I am trying to reproduce your work, while I failed to achieve the remarkable result as in your paper(current nme stuck around 0.055).
I used pdb + augment(box jitter, gaussian, rotate) + CNN6 + wing loss.

**Comment by FengZhenhua** [comment]
@alexcruz0202 Thank you. We do have pre-trained caffe and tensorflow models, converted from matconvnet. Both are working fine. We may release a pre-trained caffe model later. But I cannot guarantee the time.

**Comment by FengZhenhua** [comment]
@eyiztan Could you please provide more details for your question? For example, which dataset did you use? It is really hard to answer if you do not provide enough implementation details.

**Comment by eyiztan** [comment]
> @eyiztan Could you please provide more details for your question? For example, which dataset did you use? It is really hard to answer if you do not provide enough implementation details.

Really thanks for your quick response!
the detailed info is as following:
1. dataset: 300W, using helen, lfpw trainset and afw as training set, using helen and lfpw testset as common set, using ibug as challenge set
2. bounding box: **using ground truth box**, (we use ground truth box to avoid error imported by inaccurate face detection) ,  make the bounding box square by set the width and height to max(width, height), and we did not do any box extension.  During the evaluation and test part, we also use the ground truth box
3. then do PDB
4. 50% sample **do left, right flip**(box and shape flip in the same way),    then  do  [−30, 30] degrees  random  image  rotate(shape and box rotate in the same way), then do box jitter with  [-5% , 5%] * box_size ,  then  50% sample do gaussian blur with delta set to 1
crop the image as the box size and resize to  64*64
**we also  normalized the pixel value to [-1, 1]** 
5. choose CNN6 as the net arch, using **tf.contrib.layers.variance_scaling_initializer() as weights_initializer** 
6. using wing loss with w=10 delta =2,  using l2 regularization, with weight decay set to 5e-4
7. using learning rate set to 1e-4(3e-5 mentioned in paper was also used before), using staircase(step set to 1000)  exponential_decay with decay_rate set to 0.95.
8. batch size set to 32 (larger than paper)
9. do 120k iterations
10. we get nme(normalized by inter-pupil distance )  0.55 for common dataset, 0.10 for chanllenge dataset.   here, inter-pupil distance was calculated by mean point of right and left eye

did I miss something important?   like, normalize the label?

**Comment by FengZhenhua** [comment]
@eyiztan Thank you for providing those details and sorry for the late reply. But I have got many deadlines recently and I am really busy now. I will check your details later and response you as soon as possible. Thank you for understanding.

**Comment by eyiztan** [comment]
@FengZhenhua  Great to know you are still following the issue! let me keep expecting

**Comment by FengZhenhua** [comment]
@eyiztan For your implementation:
1. I think the dataset setting is fine.

2. Bounding box: Because you are using ground truth bounding box, it should give you better performance. But your performance is only 0.055, which is weird. When you set the bbox to be squared. did you just change the value of W/H, or you also adjust (x,y)? I suggest you adjust x/y when you reset w/h such that the face is at the middle of the image.

3. PDB should help. Do you have the results with/without PDB? Did PDB help?

4. I am not quite clear how did you implement the data augmentation. I do random augmentation for each training sample in each mini-batch. For example, for each training image, it gets 50% probability to be left-right flipped. If you run your training for 100 epochs, then all the training images should be flipped around 50 times. But it seems that you only flipped 50% images? Data augmentation is quite important and please do carefully double check this part, especially when you rotate an image, please check the rotated landmarks, this may cause some bias/errors in your training data.

5. Should be fine.

6. Wing loss looks fine. 1e-4 might be better for weight decay. 

7. I reduce the learning rate from 3e-5 to 3e-7 linearly for all the epochs.

8+9. If you use a bigger batch size, please try to train the model with more epochs. 

10. This result looks very bad.

You can normalise the label to [0,1], but then you have to change the wing loss parameters and learning rate as well. I used a label in [0,64] for CNN6 and I did not normalise the wing loss by the number of batch size. Because you are using 32 as your batch size, you should reduce your learning rate. But if you normalised the wing loss by the batch size, you should increase the learning rate.

**Comment by alexcruz0202** [comment]
@FengZhenhua.  Thank you for your detail information. Before that, I have used MSE loss and normalised the label to [0, 1]. But landmark results for closed eye wasn't successful. 
Have you ever been used this way? When I use the wing loss instead of MSE loss, I want to know parameters for it details. I haven't still depth study for your paper.  Regard.

**Comment by JoyHuYY1412** [comment]
> > @eyiztan Could you please provide more details for your question? For example, which dataset did you use? It is really hard to answer if you do not provide enough implementation details.
> > @FengZhenhua
> > Really thanks for your quick response!
> > the detailed info is as following:
> 
> 1. dataset: 300W, using helen, lfpw trainset and afw as training set, using helen and lfpw testset as common set, using ibug as challenge set
> 2. bounding box: **using ground truth box**, (we use ground truth box to avoid error imported by inaccurate face detection) ,  make the bounding box square by set the width and height to max(width, height), and **we did not do any box extension**.  During the evaluation and test part, we also use the ground truth box
> 3. then do PDB
> 4. 50% sample **do left, right flip**(box and shape flip in the same way),    then  do  [−30, 30] degrees  random  image  rotate(shape and box rotate in the same way), then do box jitter with  [-5% , 5%] * box_size ,  then  50% sample do gaussian blur with delta set to 1
>    crop the image as the box size and resize to  64*64
>    **we also  normalized the pixel value to [-1, 1]**
> 5. choose CNN6 as the net arch, using **tf.contrib.layers.variance_scaling_initializer() as weights_initializer**
> 6. using wing loss with w=10 delta =2,  using l2 regularization, with weight decay set to 5e-4
> 7. using learning rate set to 1e-4(3e-5 mentioned in paper was also used before), using staircase(step set to 1000)  exponential_decay with decay_rate set to 0.95.
> 8. batch size set to 32 (larger than paper)
> 9. do 120k iterations
> 10. we get nme(normalized by inter-pupil distance )  0.055 for common dataset, 0.10 for chanllenge dataset.   here, inter-pupil distance was calculated by mean point of right and left eye
> 
> did I miss something important? like, normalize the label?

Hi, could you please share your code for box augmentation?
